packages/dds/tree/src/feature-libraries/modular-schema/modularChangeCodecs.ts

```diff
@@ -321,31 +319,41 @@ function makeModularChangeCodec(
 	}
 
 	function encodeDetachedNodes(
-		detachedNodes: ChangeAtomIdMap<TreeChunk> | undefined,
+		detachedNodes: ChangeAtomIdBTree<TreeChunk> | undefined,
 		context: ChangeEncodingContext,
 	): EncodedBuilds | undefined {
 		if (detachedNodes === undefined) {
 			return undefined;
 		}
 
 		const treesToEncode: ITreeCursorSynchronous[] = [];
-		const buildsArray: EncodedBuildsArray = Array.from(detachedNodes.entries()).map(
-			([r, commitBuilds]) => {
-				const commitBuildsEncoded: [ChangesetLocalId, number][] = Array.from(
-					commitBuilds.entries(),
-				).map(([i, t]) => {
-					treesToEncode.push(t.cursor());
-					const treeIndexInBatch = treesToEncode.length - 1;
-					return [i, treeIndexInBatch];
-				});
-				// `undefined` does not round-trip through JSON strings, so it needs special handling.
-				// Most entries will have an undefined revision due to the revision information being inherited from the `ModularChangeset`.
-				// We therefore optimize for the common case by omitting the revision when it is undefined.
-				return r === undefined || r === context.revision
-					? [commitBuildsEncoded]
-					: [commitBuildsEncoded, revisionTagCodec.encode(r, context)];
-			},
-		);
+		const buildsArray: EncodedBuildsArray = [];
+
+		let buildsForRevision:
+			| [[ChangesetLocalId, number][], EncodedRevisionTag]
+			| [[ChangesetLocalId, number][]]
+			| undefined;
+
+		for (const [[revision, id], chunk] of detachedNodes.entries()) {
+			const encodedRevision = encodeRevisionOpt(revisionTagCodec, revision, context);
+
+			if (buildsForRevision === undefined || buildsForRevision[1] !== encodedRevision) {
+				if (buildsForRevision !== undefined) {
+					buildsArray.push(buildsForRevision);
+				}
+
+				buildsForRevision = encodedRevision !== undefined ? [[], encodedRevision] : [[]];
+			}
+
+			treesToEncode.push(chunk.cursor());
+			const treeIndexInBatch = treesToEncode.length - 1;
+			buildsForRevision?.[0].push([id, treeIndexInBatch]);
+		}
+
+		if (buildsForRevision !== undefined) {
+			buildsArray.push(buildsForRevision);
+		}
+
```

Comment:
It seems that this code will at most add `buildsForRevision` to `buildsArray` a single time. That seems like a bug since we want to support multiple revisions.
Would be good to add a test that to catch this kind of regression.

Reply:
I can see if I can write this in an easier to follow way, but we also push to buildsArray on line 342. I think I did see a test that encodes multiple build revisions.

Reply:
Oh dang, I missed that. I can see how if adverts the issue I was thinking of.
This code is indeed pretty ugly (which predates this PR), up to you if you feel like making it better right now or not.